pydoctor/test/test_epydoc2stan.py

```diff
@@ -856,3 +856,77 @@ def test_annotation_formatter(annotation: str) -> None:
     assert html.endswith('</code>')
     text = html[6:-7]
     assert text.replace('<wbr></wbr>', '').replace('<wbr>\n</wbr>', '') == expected_text
+
+def test_module_docformat(capsys: CapSys) -> None:
+    """
+    Test if Module.docformat effectively override System.options.docformat
+    """
+
+    system = model.System()
+    system.options.docformat = 'restructuredtext'
+
+    mod = fromText('''
+    """
+    Link to pydoctor: U{pydoctor <https://github.com/twisted/pydoctor>}.
+    """
+    __docformat__ = "epytext"
+    ''', modname='test_epy', system=system)
+
+    epytext_output = epydoc2stan.format_docstring(mod)
+
+    captured = capsys.readouterr().out
+    assert not captured
+
+    mod = fromText('''
+    """
+    Link to pydoctor: `pydoctor <https://github.com/twisted/pydoctor>`_.
+    """
+    __docformat__ = "restructuredtext en"
+    ''', modname='test_rst', system=system)
+
+    restructuredtext_output = epydoc2stan.format_docstring(mod)
+
+    captured = capsys.readouterr().out
+    assert not captured
+
+    assert ('Link to pydoctor: <a href="https://github.com/twisted/pydoctor"'
+        ' target="_top">pydoctor</a>' in flatten(epytext_output))
+    
+    assert ('Link to pydoctor: <a class="rst-reference external"'
+        ' href="https://github.com/twisted/pydoctor" target="_top">pydoctor</a>' in flatten(restructuredtext_output))
+
+
+    top_src = '''
+    def f(a: str, b: int): 
+        """
+        :param a: string
+        :param b: integer
+        """
+        pass
+    '''
+    mod_src = '''
+    def f(a: str, b: int): 
+        """
+        @param a: string
+        @param b: integer
+        """
+        pass
+    '''
+    pkg_src = '''
+    __docformat__ = 'epytext'
+    '''
+
+    system = model.System()
+    system.options.docformat = 'restructuredtext'
+    top = fromText(top_src, modname='top', is_package=True, system=system)
+    fromText(pkg_src, modname='pkg', parent_name='top', is_package=True,
+                   system=system)
+    mod = fromText(mod_src, modname='top.pkg.mod', parent_name='top.pkg', system=system)
+    
+    captured = capsys.readouterr().out
+    assert not captured
+
+    assert ''.join(docstring2html(top.contents['f']).splitlines()) == ''.join(docstring2html(mod.contents['f']).splitlines())
+    
```

Comment:
This is the test for the undocumented format inheritance support, correct? I think it would be useful to first discuss if that is something we want, and if so, how exactly it should work and document that.

Currently it looks only one level up, which seems weird to me: if we'd want to inherit the docstring format, I think we'd want check all the way up to the top-level package. Perhaps we'd even want the top-level package to default to the `System` value then, instead of having `get_parser()` do that.

Reply:
> > Currently it looks only one level up
> 
> No, it does lookup the whole hierarchy of packages, I think. I'll add a test to make sure of that.

Ah, indeed it does. I was confused by this line of code:
```py
_docformat = getattr(self.parent, "docformat", None)
```
This reads the parent property, but because of the variable name I thought it read the instance variable instead. So please change the variable name.

> > Perhaps we'd even want the top-level package to default to the System value then, instead of having get_parser() do that.
> 
> I see that as more complication honestly

It depends on how you look at it: it you see the property as "give me the docstring format override for this module", then it's a complication. But if you see it as "give me the docstring format that applies to this module", then it's actually simpler, since the whole concept of overrides is no longer a concern for the caller.

Reply:
Well, I guess it depends how we see it. I went with "give me the docstring format override for this module" because this is also the wish of the user when using this feature. But the other way around is also justifiable.

Reply:
I always prefer to get rid of `Optional` when possible. But it's not a big deal; let's leave it like this for now.